Packages/com.unity.inputsystem/InputSystem/Plugins/UI/UIInputModule.cs

```diff
@@ -16,30 +16,360 @@ namespace UnityEngine.Experimental.Input.Plugins.UI
+
+            MouseModel.InternalData internalMouseData = mouseState.internalData;
+            internalMouseData.hoverTargets = eventData.hovered;
+            internalMouseData.pointerTarget = eventData.pointerEnter;
+            mouseState.internalData = internalMouseData;
+
+            ProcessMouseButtonDrag(eventData);
+
+            buttonState.CopyFrom(eventData);
+            mouseState.leftButton = buttonState;
+
+            /// Right Mouse Button
+            buttonState = mouseState.rightButton;
+            buttonState.CopyTo(eventData);
+
+            ProcessMouseButton(buttonState.lastFrameDelta, eventData);
+            ProcessMouseButtonDrag(eventData);
+
+            buttonState.CopyFrom(eventData);
+            mouseState.rightButton = buttonState;
+
+            /// Middle Mouse Button
+            buttonState = mouseState.middleButton;
+            buttonState.CopyTo(eventData);
+
+            ProcessMouseButton(buttonState.lastFrameDelta, eventData);
+            ProcessMouseButtonDrag(eventData);
+
+            buttonState.CopyFrom(eventData);
+            mouseState.middleButton = buttonState;
+
+            mouseState.OnFrameFinished();
         }
 
-        protected void PerformRaycast(PointerEventData eventData)
+        private void ProcessMouseMovement(PointerEventData eventData)
         {
-            if (eventData == null)
-                throw new ArgumentNullException("eventData");
+            // walk up the tree till a common root between the last entered and the current entered is foung
+            // send exit events up to (but not inluding) the common root. Then send enter events up to
+            // (but not including the common root).
 
-            eventSystem.RaycastAll(eventData, m_RaycastResultCache);
-            eventData.pointerCurrentRaycast = FindFirstRaycast(m_RaycastResultCache);
-            m_RaycastResultCache.Clear();
+            GameObject currentPointerTarget = eventData.pointerCurrentRaycast.gameObject;
+
+            // if we have no target / pointerEnter has been deleted
+            // just send exit events to anything we are tracking
+            // then exit
+            if (currentPointerTarget == null || eventData.pointerEnter == null)
+            {
+                for (var i = 0; i < eventData.hovered.Count; ++i)
+                    ExecuteEvents.Execute(eventData.hovered[i], eventData, ExecuteEvents.pointerExitHandler);
+
+                eventData.hovered.Clear();
+
+                if (currentPointerTarget == null)
+                {
+                    eventData.pointerEnter = null;
+                    return;
+                }
+            }
+
+            // if we have not changed hover target
+            if (eventData.pointerEnter == currentPointerTarget && currentPointerTarget)
+                return;
+
+            var commonRoot = FindCommonRoot(eventData.pointerEnter, currentPointerTarget);
+
+            // and we already an entered object from last time
```

Comment:
Minor thing but please use proper English orthography in comments. Full sentences starting upper and ended with a dot.